src/backend/columnar/cstore_tableam.c

```diff
@@ -1289,6 +1209,78 @@ CStoreTableAMObjectAccessHook(ObjectAccessType access, Oid classId, Oid objectId
 }
 
 
+/*
+ * Reject AFTER ... FOR EACH ROW triggers on columnar tables.
+ */
+static void
+CStoreTriggerCreateHook(Oid tgid)
+{
+	/*
+	 * Fetch the pg_trigger tuple by the Oid of the trigger
+	 */
+	ScanKeyData skey[1];
+	Relation tgrel = table_open(TriggerRelationId, AccessShareLock);
+
+	ScanKeyInit(&skey[0],
+				Anum_pg_trigger_oid,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(tgid));
+
+	SysScanDesc tgscan = systable_beginscan(tgrel, TriggerOidIndexId, true,
+											SnapshotSelf, 1, skey);
+
+	HeapTuple tgtup = systable_getnext(tgscan);
+
+	if (!HeapTupleIsValid(tgtup))
+	{
+		table_close(tgrel, AccessShareLock);
+		return;
+	}
+
+	Form_pg_trigger tgrec = (Form_pg_trigger) GETSTRUCT(tgtup);
+
+	Oid tgrelid = tgrec->tgrelid;
+	int16 tgtype = tgrec->tgtype;
+
+	systable_endscan(tgscan);
+	table_close(tgrel, AccessShareLock);
+
+	if (TRIGGER_FOR_ROW(tgtype) && TRIGGER_FOR_AFTER(tgtype) &&
+		IsCStoreTableAmTable(tgrelid))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg(
+							"Foreign keys and AFTER ROW triggers are not supported for columnar tables"),
+						errhint("Consider an AFTER STATEMENT trigger instead.")));
+	}
+}
+
+
+/*
+ * Capture create/drop events and dispatch to the proper action.
+ */
+static void
+CStoreTableAMObjectAccessHook(ObjectAccessType access, Oid classId, Oid objectId, int
+							  subId,
+							  void *arg)
```

Comment:
nit; We should really figure out how to make our formatter not split between type and variable identifiers, but I would prefer to keep them together like:
```suggestion
CStoreTableAMObjectAccessHook(ObjectAccessType access, Oid classId, Oid objectId,
							  int subId, void *arg)
```

Reply:
Fixed.